File: frontend/src/stores/raffles.ts
/**
 * Raffles store: public browsing + sign-up, and admin management (CRUD,
 * entries, winner picking, image upload). Mirrors all raffle logic from app.js.
 */
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { endpoints } from '@/lib/endpoints'
import { utcToDatetimeLocal, datetimeLocalToUtc, parseServerTimestamp } from '@/lib/datetime'
import type { Raffle, RaffleEnterResponse, RaffleEntry, RaffleForm } from '@/types/api'
import { useUiStore } from './ui'
import { useImagesStore, IMAGE_DIR_RAFFLES } from './images'
import { withLoading } from '@/lib/withLoading'

/**
 * Whether a raffle is enterable by the public right now: it must be `open` and
 * inside its availability window — past `available_from` (if set) and before
 * `available_to` (if set). This mirrors the backend's public list query
 * (store.ListRaffles, non-admin) so the two never disagree.
 *
 * The public `GET /api/raffles` already applies the same window for normal
 * visitors, so this is a second line of defence for the two cases the list
 * query can't cover: a raffle reached by a direct link to its detail page
 * (GetRaffle ignores the window), and an admin browsing the public pages (the
 * list endpoint returns every raffle in admin mode). A raffle keeps its `open`
 * status outside its window, so checking `status` alone is not enough. (Admin
 * tabs intentionally do not use this — admins manage out-of-window raffles.)
 */
export function isRaffleEnterable(r: Raffle): boolean {
  if (r.status !== 'open') return false
  const now = Date.now()
  const startsAt = parseServerTimestamp(r.available_from)
  if (!Number.isNaN(startsAt) && startsAt > now) return false // not yet open
  const endsAt = parseServerTimestamp(r.available_to)
  if (!Number.isNaN(endsAt) && endsAt <= now) return false // already ended
  return true
}

export const useRafflesStore = defineStore('raffles', () => {
  const ui = useUiStore()

  const homeRaffles = ref<Raffle[]>([]) // open raffles for home card visibility
  const raffles = ref<Raffle[]>([])
  const selectedRaffle = ref<Raffle | null>(null)
  const raffleEntries = ref<RaffleEntry[]>([])
  const raffleForm = ref<RaffleForm | null>(null)
  const raffleSignup = ref<{ characterName: string; world: string; numEntries: number }>({
    characterName: '',
    world: '',
    numEntries: 1,
  })
  const raffleSignupResult = ref<RaffleEnterResponse | null>(null)
  /** Cloudflare Turnstile token for the public sign-up (empty when disabled/unset). */
  const signupTurnstileToken = ref('')
  // Admin-only: manually add a player to the selected open raffle.
  const entryAdd = ref<{ characterName: string; world: string; numEntries: number; paid: boolean }>(
    { characterName: '', world: '', numEntries: 1, paid: false },
  )
  const addingEntry = ref(false)
  const raffleWinner = ref<RaffleEntry | null>(null)
  const raffleWinnerEntry = ref<RaffleEntry | null>(null) // public closed view
  const raffleTotalEntryCount = ref(0)
  /** Reusable prize-image paths (the "Raffle" category on the Images page). */
  const prizeImages = ref<string[]>([])
  // In-flight flags driving spinners / button disabling.
  const rafflesLoading = ref(false)
  const detailLoading = ref(false)
  // Monotonic token guarding loadRaffleDetail against a last-write-wins race:
  // two rapid opens each fire a request, and a slow earlier one could otherwise
  // overwrite a newer one. Only the latest request applies its result.
  let detailSeq = 0
  const savingRaffle = ref(false)
  const entering = ref(false)
  const pickingWinner = ref(false)

  // ── Computed ───────────────────────────────────────────────────────────────

  const openRaffles = computed(() => raffles.value.filter((r) => r.status === 'open'))
  const closedRaffles = computed(() => raffles.value.filter((r) => r.status === 'closed'))

  /** Public: is the currently-viewed raffle still enterable (open + not ended)? */
  const selectedRaffleEnterable = computed(
    () => selectedRaffle.value !== null && isRaffleEnterable(selectedRaffle.value),
  )

  // ── Load ─────────────────────────────────────────────────────────────────

  async function loadRaffles(): Promise<void> {
    await withLoading(rafflesLoading, async () => {
      const data = await endpoints.raffles.list()
      raffles.value = data.raffles
    })
  }

  /** Preloads open raffles (home page card visibility). */
  async function loadHomeRaffles(): Promise<void> {
    try {
      const data = await endpoints.raffles.list()
      homeRaffles.value = data.raffles.filter(isRaffleEnterable)
    } catch {
      /* silent */
    }
  }

  async function loadRaffleDetail(id: number): Promise<void> {
    const reqId = ++detailSeq
    detailLoading.value = true
    try {
      const data = await endpoints.raffles.detail(id)
      if (reqId !== detailSeq) return // a newer load superseded this one
      selectedRaffle.value = data.raffle
      raffleEntries.value = data.entries || []
      raffleWinner.value = null
      if (data.raffle.winner_entry_id && raffleEntries.value.length) {
        raffleWinner.value =
          raffleEntries.value.find((e) => e.id === data.raffle.winner_entry_id) || null
      }
    } catch (e) {
      if (reqId === detailSeq) ui.notify((e as Error).message, 'error')
    } finally {
      if (reqId === detailSeq) detailLoading.value = false
    }
  }

  /** Admin: open a raffle's detail view. */
  function viewRaffle(raffle: Raffle): void {
    selectedRaffle.value = raffle
    raffleSignup.value = { characterName: '', world: '', numEntries: 1 }
    raffleSignupResult.value = null
    raffleEntries.value = []
    raffleWinner.value = null
    resetEntryAdd()
    void loadRaffleDetail(raffle.id)
  }

  /** Clears the admin "add entry" form. */
  function resetEntryAdd(): void {
    entryAdd.value = { characterName: '', world: '', numEntries: 1, paid: false }
  }

  /** Public: open a raffle's detail view (loads winner + total entries). */
  function viewPublicRaffle(raffle: Raffle): void {
    selectedRaffle.value = raffle
    raffleSignup.value = { characterName: '', world: '', numEntries: 1 }
    raffleSignupResult.value = null
    raffleWinnerEntry.value = null
    raffleTotalEntryCount.value = 0
    endpoints.raffles
      .detail(raffle.id)
      .then((data) => {
        selectedRaffle.value = data.raffle
        raffleTotalEntryCount.value = data.total_entries || 0
        if (data.winner_entry) raffleWinnerEntry.value = data.winner_entry
      })
      .catch(() => {})
  }

  /**
   * Public: load a raffle's detail by id (used when navigating directly to
   * /raffles/:id, e.g. on refresh or a shared link). Resets the sign-up state,
   * then fetches the raffle + winner + total entry count. Returns true on
   * success, false on failure (so the view can redirect back to the list).
   */
  async function loadPublicRaffleById(id: number): Promise<boolean> {
    selectedRaffle.value = null
    raffleSignup.value = { characterName: '', world: '', numEntries: 1 }
    raffleSignupResult.value = null
    raffleWinnerEntry.value = null
    raffleTotalEntryCount.value = 0
    detailLoading.value = true
    try {
      const data = await endpoints.raffles.detail(id)
      selectedRaffle.value = data.raffle
      raffleTotalEntryCount.value = data.total_entries || 0
      if (data.winner_entry) raffleWinnerEntry.value = data.winner_entry
      return true
    } catch {
      return false
    } finally {
      detailLoading.value = false
    }
  }

  // ── Admin form ─────────────────────────────────────────────────────────────

  function newRaffleForm(): void {
    raffleForm.value = {
      id: 0,
      title: '',
      description: '',
      rules: '',
      max_entries: 1,
      signup_instructions: '',
      cost_per_entry: 0,
      available_from: '',
      available_to: '',
      prize_image: '',
    }
  }

  function editRaffleForm(raffle: Raffle): void {
    // Availability dates are stored as UTC; convert to local time so the
    // datetime-local inputs show the correct wall-clock for *this* admin's
    // timezone (a window set by an admin in another zone reads correctly).
    raffleForm.value = {
      ...(raffle as unknown as RaffleForm),
      available_from: utcToDatetimeLocal(raffle.available_from),
      available_to: utcToDatetimeLocal(raffle.available_to),
    }
  }

  /**
   * Seed a brand-new raffle form from an existing (e.g. closed) raffle — copies
   * the reusable content (title, markdown bodies, limits, cost, prize image) but
   * starts with a cleared availability window and a zero id, so saving creates a
   * fresh open raffle rather than editing the original.
   */
  function copyRaffleForm(raffle: Raffle): void {
    raffleForm.value = {
      id: 0,
      title: raffle.title,
      description: raffle.description,
      rules: raffle.rules,
      max_entries: raffle.max_entries,
      signup_instructions: raffle.signup_instructions,
      cost_per_entry: raffle.cost_per_entry,
      available_from: '',
      available_to: '',
      prize_image: raffle.prize_image,
    }
  }

  function cancelRaffleForm(): void {
    raffleForm.value = null
  }

  /** Saves the raffle form. Returns true on success (caller navigates). */
  async function saveRaffle(): Promise<boolean> {
    if (!raffleForm.value) return false
    const f = raffleForm.value
    if (!f.title.trim()) {
      ui.notify('Title is required', 'error')
      return false
    }
    savingRaffle.value = true
    try {
      // The form holds local datetime-local values; convert the availability
      // window to UTC so the stored instant is timezone-unambiguous.
      const payload = {
        ...f,
        available_from: datetimeLocalToUtc(f.available_from),
        available_to: datetimeLocalToUtc(f.available_to),
      }
      if (f.id) {
        await endpoints.raffles.update(f.id, payload)
        ui.notify('Raffle updated', 'success')
      } else {
        await endpoints.raffles.create(payload)
        ui.notify('Raffle created', 'success')
      }
      raffleForm.value = null
      await loadRaffles()
      return true
    } catch (e) {
      ui.notify((e as Error).message, 'error')
      return false
    } finally {
      savingRaffle.value = false
    }
  }

  async function deleteRaffle(id: number): Promise<void> {
    if (
      !(await ui.confirm('Delete this raffle and all its entries?', {
        title: 'Delete raffle',
        confirmText: 'Delete',
      }))
    )
      return
    try {
      await endpoints.raffles.delete(id)
      raffles.value = raffles.value.filter((r) => r.id !== id)
      if (selectedRaffle.value && selectedRaffle.value.id === id) selectedRaffle.value = null
      ui.notify('Raffle deleted', 'info')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  /** Loads the reusable prize images (the "Raffle" category) for the form picker.
   *  Prize images are stored as root-relative paths, so the picker uses `.path`. */
  async function loadPrizeImages(): Promise<void> {
    try {
      const images = useImagesStore()
      await images.loadImages(IMAGE_DIR_RAFFLES)
      prizeImages.value = images.imagesByDir[IMAGE_DIR_RAFFLES].map((i) => i.path)
    } catch {
      /* non-fatal: the picker just shows nothing */
    }
  }

  // ── Public sign-up ─────────────────────────────────────────────────────────

  /**
   * The entry count clamped to a whole number in [1, max_entries]. The raw
   * `numEntries` can be out of range or non-integer (typed value, stepper, or a
   * cleared field), so both the live cost preview and the submitted request go
   * through this — the displayed total can never disagree with what's sent, and
   * the server's own bound is never the first line of defence.
   */
  function clampedEntries(): number {
    if (!selectedRaffle.value) return 1
    const max = Math.max(1, Math.floor(selectedRaffle.value.max_entries || 1))
    const raw = Math.floor(raffleSignup.value.numEntries || 1)
    return Math.min(Math.max(raw, 1), max)
  }

  /** Writes the clamped entry count back to the field (call on input blur/change). */
  function clampSignupEntries(): void {
    raffleSignup.value.numEntries = clampedEntries()
  }

  function raffleTotalCost(): number {
    if (!selectedRaffle.value) return 0
    return clampedEntries() * selectedRaffle.value.cost_per_entry
  }

  async function enterRaffle(): Promise<void> {
    if (!selectedRaffle.value) return
    const s = raffleSignup.value
    if (!s.characterName.trim() || !s.world.trim()) {
      ui.notify('Character name and world are required', 'error')
      return
    }
    clampSignupEntries()
    entering.value = true
    try {
      const data = await endpoints.raffles.enter(selectedRaffle.value.id, {
        character_name: s.characterName.trim(),
        world: s.world.trim(),
        num_entries: s.numEntries,
        turnstile_token: signupTurnstileToken.value || undefined,
      })
      raffleSignupResult.value = data
      ui.notify(data.message, 'success')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      // Turnstile tokens are single-use — clear it so the widget re-issues one.
      signupTurnstileToken.value = ''
      entering.value = false
    }
  }

  // ── Admin entries ──────────────────────────────────────────────────────────

  /**
   * Admin: manually add a player to the selected open raffle (optionally already
   * paid). The entry count is clamped to [1, max_entries] to match the field's
   * bound; the server enforces the same limit. On success the form resets and the
   * detail view reloads so the entries table + counts reflect the new entry.
   */
  async function addRaffleEntry(): Promise<void> {
    if (!selectedRaffle.value) return
    const f = entryAdd.value
    if (!f.characterName.trim() || !f.world.trim()) {
      ui.notify('Character name and world are required', 'error')
      return
    }
    const max = Math.max(1, Math.floor(selectedRaffle.value.max_entries || 1))
    const num = Math.min(Math.max(Math.floor(f.numEntries || 1), 1), max)
    addingEntry.value = true
    try {
      await endpoints.raffles.addEntry(selectedRaffle.value.id, {
        character_name: f.characterName.trim(),
        world: f.world.trim(),
        num_entries: num,
        paid: f.paid,
      })
      ui.notify('Entry added', 'success')
      resetEntryAdd()
      await loadRaffleDetail(selectedRaffle.value.id)
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      addingEntry.value = false
    }
  }

  async function toggleEntryPaid(entry: RaffleEntry): Promise<void> {
    if (!selectedRaffle.value) return
    try {
      await endpoints.raffles.markEntryPaid(selectedRaffle.value.id, entry.id, !entry.paid)
      entry.paid = !entry.paid
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  async function deleteEntry(entry: RaffleEntry): Promise<void> {
    if (!selectedRaffle.value) return
    if (!(await ui.confirm('Delete this entry?', { title: 'Delete entry', confirmText: 'Delete' })))
      return
    try {
      await endpoints.raffles.deleteEntry(selectedRaffle.value.id, entry.id)
      raffleEntries.value = raffleEntries.value.filter((e) => e.id !== entry.id)
      ui.notify('Entry deleted', 'info')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  async function pickRaffleWinner(): Promise<void> {
    if (!selectedRaffle.value) return
    pickingWinner.value = true
    try {
      const data = await endpoints.raffles.pickWinner(selectedRaffle.value.id)
      raffleWinner.value = data.winner
      selectedRaffle.value.winner_entry_id = data.winner.id
      ui.notify('Winner picked!', 'success')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      pickingWinner.value = false
    }
  }

  async function verifyRaffleWinner(): Promise<void> {
    if (!selectedRaffle.value) return
    try {
      await endpoints.raffles.verifyWinner(selectedRaffle.value.id)
      selectedRaffle.value.status = 'closed'
      ui.notify('Winner verified! Raffle closed.', 'success')
      await loadRaffles()
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  async function pickAnotherWinner(): Promise<void> {
    if (!selectedRaffle.value) return
    pickingWinner.value = true
    try {
      const data = await endpoints.raffles.pickAnotherWinner(selectedRaffle.value.id)
      raffleWinner.value = data.winner
      selectedRaffle.value.winner_entry_id = data.winner.id
      ui.notify('New winner picked!', 'success')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      pickingWinner.value = false
    }
  }

  return {
    homeRaffles,
    raffles,
    selectedRaffle,
    raffleEntries,
    raffleForm,
    raffleSignup,
    raffleSignupResult,
    signupTurnstileToken,
    entryAdd,
    addingEntry,
    raffleWinner,
    raffleWinnerEntry,
    raffleTotalEntryCount,
    prizeImages,
    rafflesLoading,
    detailLoading,
    savingRaffle,
    entering,
    pickingWinner,
    openRaffles,
    closedRaffles,
    selectedRaffleEnterable,
    loadRaffles,
    loadHomeRaffles,
    loadRaffleDetail,
    viewRaffle,
    viewPublicRaffle,
    loadPublicRaffleById,
    newRaffleForm,
    editRaffleForm,
    copyRaffleForm,
    cancelRaffleForm,
    saveRaffle,
    deleteRaffle,
    loadPrizeImages,
    raffleTotalCost,
    clampSignupEntries,
    enterRaffle,
    addRaffleEntry,
    toggleEntryPaid,
    deleteEntry,
    pickRaffleWinner,
    verifyRaffleWinner,
    pickAnotherWinner,
  }
})

File: backend/internal/store/raffles.go
package store

import (
	"database/sql"
	"errors"

	"app-suite/internal/model"
)

// ── Raffle operations ───────────────────────────────────────────────────────

// CreateRaffle inserts a new raffle and returns its ID.
func (s *Store) CreateRaffle(r *model.Raffle) (int64, error) {
	res, err := s.db.Exec(`INSERT INTO raffles (title, description, rules, max_entries, signup_instructions, cost_per_entry, available_from, available_to, prize_image, status)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		r.Title, r.Description, r.Rules, r.MaxEntries, r.SignupInstructions,
		r.CostPerEntry, r.AvailableFrom, r.AvailableTo, r.PrizeImage, "open",
	)
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

// UpdateRaffle updates an existing raffle's editable fields.
func (s *Store) UpdateRaffle(r *model.Raffle) error {
	_, err := s.db.Exec(`UPDATE raffles SET title=?, description=?, rules=?, max_entries=?, signup_instructions=?, cost_per_entry=?, available_from=?, available_to=?, prize_image=? WHERE id=?`,
		r.Title, r.Description, r.Rules, r.MaxEntries, r.SignupInstructions,
		r.CostPerEntry, r.AvailableFrom, r.AvailableTo, r.PrizeImage, r.ID,
	)
	return err
}

// DeleteRaffle removes a raffle and its entries (cascade). Returns true if deleted.
func (s *Store) DeleteRaffle(id int64) (bool, error) {
	res, err := s.db.Exec("DELETE FROM raffles WHERE id = ?", id)
	if err != nil {
		return false, err
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}

// GetRaffle retrieves a single raffle by ID. Returns nil if not found.
func (s *Store) GetRaffle(id int64) (*model.Raffle, error) {
	var r model.Raffle
	var winnerID sql.NullInt64
	err := s.db.QueryRow(`SELECT id, title, description, rules, max_entries, signup_instructions, cost_per_entry, available_from, available_to, prize_image, status, winner_entry_id, created_at FROM raffles WHERE id = ?`, id).
		Scan(&r.ID, &r.Title, &r.Description, &r.Rules, &r.MaxEntries, &r.SignupInstructions,
			&r.CostPerEntry, &r.AvailableFrom, &r.AvailableTo, &r.PrizeImage, &r.Status, &winnerID, &r.CreatedAt)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	if winnerID.Valid {
		r.WinnerEntryID = &winnerID.Int64
	}
	return &r, nil
}

// ListRaffles returns raffles. If adminMode is false, returns open raffles within the
// availability date range plus all closed raffles (for browsing past results).
//
// In admin mode each raffle also carries two read-only aggregates used by the
// closed-raffle table: WinnerName (the verified winner entry's "Character @ World",
// joined from raffle_entries) and PaidTotal (the sum of paid tickets × cost_per_entry,
// the gil collected). The public list omits both.
func (s *Store) ListRaffles(adminMode bool) ([]model.Raffle, error) {
	if adminMode {
		return s.listRafflesAdmin()
	}
	// Public list: only open raffles currently inside their availability
	// window. Availability dates are stored as UTC (RFC-3339 with 'Z' for new
	// values, legacy naive strings treated as UTC); datetime() normalizes both
	// to a UTC timestamp so the comparison against datetime('now') (also UTC)
	// is timezone-correct — a raffle past its "available to" instant no longer
	// shows regardless of the admin's timezone.
	rows, err := s.db.Query(`SELECT id, title, description, rules, max_entries, signup_instructions, cost_per_entry, available_from, available_to, prize_image, status, winner_entry_id, created_at FROM raffles WHERE status = 'open' AND (available_from = '' OR datetime(available_from) <= datetime('now')) AND (available_to = '' OR datetime(available_to) >= datetime('now')) ORDER BY created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	raffles := make([]model.Raffle, 0)
	for rows.Next() {
		var r model.Raffle
		var winnerID sql.NullInt64
		if err := rows.Scan(&r.ID, &r.Title, &r.Description, &r.Rules, &r.MaxEntries, &r.SignupInstructions,
			&r.CostPerEntry, &r.AvailableFrom, &r.AvailableTo, &r.PrizeImage, &r.Status, &winnerID, &r.CreatedAt); err != nil {
			return nil, err
		}
		if winnerID.Valid {
			r.WinnerEntryID = &winnerID.Int64
		}
		raffles = append(raffles, r)
	}
	return raffles, rows.Err()
}

// listRafflesAdmin returns every raffle with the closed-table aggregates joined in:
// the winner entry's name (when a winner is set) and the total gil collected from
// paid tickets. The paid-ticket sum comes from a correlated subquery so raffles
// with no paid entries still report 0.
func (s *Store) listRafflesAdmin() ([]model.Raffle, error) {
	rows, err := s.db.Query(`SELECT r.id, r.title, r.description, r.rules, r.max_entries, r.signup_instructions,
			r.cost_per_entry, r.available_from, r.available_to, r.prize_image, r.status, r.winner_entry_id, r.created_at,
			COALESCE(w.character_name, ''), COALESCE(w.world, ''),
			COALESCE((SELECT SUM(e.num_entries) FROM raffle_entries e WHERE e.raffle_id = r.id AND e.paid = 1), 0)
		FROM raffles r
		LEFT JOIN raffle_entries w ON w.id = r.winner_entry_id
		ORDER BY r.created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	raffles := make([]model.Raffle, 0)
	for rows.Next() {
		var r model.Raffle
		var winnerID sql.NullInt64
		var winnerChar, winnerWorld string
		var paidTickets int
		if err := rows.Scan(&r.ID, &r.Title, &r.Description, &r.Rules, &r.MaxEntries, &r.SignupInstructions,
			&r.CostPerEntry, &r.AvailableFrom, &r.AvailableTo, &r.PrizeImage, &r.Status, &winnerID, &r.CreatedAt,
			&winnerChar, &winnerWorld, &paidTickets); err != nil {
			return nil, err
		}
		if winnerID.Valid {
			r.WinnerEntryID = &winnerID.Int64
			if winnerChar != "" {
				r.WinnerName = winnerChar + " @ " + winnerWorld
			}
		}
		r.PaidTotal = float64(paidTickets) * r.CostPerEntry
		raffles = append(raffles, r)
	}
	return raffles, rows.Err()
}

// CountRaffleEntries returns the total number of entries (sum of num_entries) for a raffle.
func (s *Store) CountRaffleEntries(raffleID int64) (int, error) {
	var count int
	err := s.db.QueryRow(`SELECT COALESCE(SUM(num_entries), 0) FROM raffle_entries WHERE raffle_id = ?`, raffleID).Scan(&count)
	return count, err
}

// ListRaffleEntries returns all entries for a raffle.
func (s *Store) ListRaffleEntries(raffleID int64) ([]model.RaffleEntry, error) {
	rows, err := s.db.Query(`SELECT id, raffle_id, character_name, world, num_entries, paid, created_at FROM raffle_entries WHERE raffle_id = ? ORDER BY created_at ASC`, raffleID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	entries := make([]model.RaffleEntry, 0)
	for rows.Next() {
		var e model.RaffleEntry
		var paid int
		if err := rows.Scan(&e.ID, &e.RaffleID, &e.CharacterName, &e.World, &e.NumEntries, &paid, &e.CreatedAt); err != nil {
			return nil, err
		}
		e.Paid = paid != 0
		entries = append(entries, e)
	}
	return entries, rows.Err()
}

// GetRaffleEntry retrieves a single raffle entry by character+world for a raffle.
func (s *Store) GetRaffleEntry(raffleID int64, charName, world string) (*model.RaffleEntry, error) {
	var e model.RaffleEntry
	var paid int
	err := s.db.QueryRow(`SELECT id, raffle_id, character_name, world, num_entries, paid, created_at FROM raffle_entries WHERE raffle_id = ? AND LOWER(character_name) = LOWER(?) AND LOWER(world) = LOWER(?)`,
		raffleID, charName, world).
		Scan(&e.ID, &e.RaffleID, &e.CharacterName, &e.World, &e.NumEntries, &paid, &e.CreatedAt)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	e.Paid = paid != 0
	return &e, nil
}

// GetRaffleEntryByID returns a single raffle entry by its ID.
func (s *Store) GetRaffleEntryByID(entryID int64) (*model.RaffleEntry, error) {
	var e model.RaffleEntry
	var paid int
	err := s.db.QueryRow(`SELECT id, raffle_id, character_name, world, num_entries, paid, created_at FROM raffle_entries WHERE id = ?`,
		entryID).
		Scan(&e.ID, &e.RaffleID, &e.CharacterName, &e.World, &e.NumEntries, &paid, &e.CreatedAt)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	e.Paid = paid != 0
	return &e, nil
}

// CreateRaffleEntry inserts a new raffle entry.
func (s *Store) CreateRaffleEntry(raffleID int64, charName, world string, numEntries int) (int64, error) {
	res, err := s.db.Exec(`INSERT INTO raffle_entries (raffle_id, character_name, world, num_entries) VALUES (?, ?, ?, ?)`,
		raffleID, charName, world, numEntries)
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

// AddRaffleEntries adds entries to an existing raffle_entries row.
func (s *Store) AddRaffleEntries(entryID int64, additionalEntries int) error {
	_, err := s.db.Exec(`UPDATE raffle_entries SET num_entries = num_entries + ? WHERE id = ?`, additionalEntries, entryID)
	return err
}

// SetRaffleEntryPaid sets the paid flag on a raffle entry.
func (s *Store) SetRaffleEntryPaid(entryID int64, paid bool) error {
	val := 0
	if paid {
		val = 1
	}
	_, err := s.db.Exec(`UPDATE raffle_entries SET paid = ? WHERE id = ?`, val, entryID)
	return err
}

// SetRaffleWinner sets the winner_entry_id on a raffle.
func (s *Store) SetRaffleWinner(raffleID int64, entryID *int64) error {
	if entryID == nil {
		_, err := s.db.Exec(`UPDATE raffles SET winner_entry_id = NULL WHERE id = ?`, raffleID)
		return err
	}
	_, err := s.db.Exec(`UPDATE raffles SET winner_entry_id = ? WHERE id = ?`, *entryID, raffleID)
	return err
}

// SetRaffleStatus updates the raffle's status.
func (s *Store) SetRaffleStatus(raffleID int64, status string) error {
	_, err := s.db.Exec(`UPDATE raffles SET status = ? WHERE id = ?`, status, raffleID)
	return err
}

// PickRaffleWinner picks a random paid entry weighted by num_entries.
// Returns nil if no paid entries exist.
func (s *Store) PickRaffleWinner(raffleID int64) (*model.RaffleEntry, error) {
	// Get all paid entries
	rows, err := s.db.Query(`SELECT id, raffle_id, character_name, world, num_entries, paid, created_at FROM raffle_entries WHERE raffle_id = ? AND paid = 1`, raffleID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var entries []model.RaffleEntry
	totalTickets := 0
	for rows.Next() {
		var e model.RaffleEntry
		var paid int
		if err := rows.Scan(&e.ID, &e.RaffleID, &e.CharacterName, &e.World, &e.NumEntries, &paid, &e.CreatedAt); err != nil {
			return nil, err
		}
		e.Paid = paid != 0
		entries = append(entries, e)
		totalTickets += e.NumEntries
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	if len(entries) == 0 || totalTickets == 0 {
		return nil, nil
	}

	// Weighted random pick
	pick := randInt(totalTickets)
	cumulative := 0
	for _, e := range entries {
		cumulative += e.NumEntries
		if pick < cumulative {
			return &e, nil
		}
	}
	// Fallback (shouldn't reach)
	return &entries[len(entries)-1], nil
}

// DeleteRaffleEntry removes a raffle entry by ID.
func (s *Store) DeleteRaffleEntry(entryID int64) (bool, error) {
	res, err := s.db.Exec("DELETE FROM raffle_entries WHERE id = ?", entryID)
	if err != nil {
		return false, err
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}
